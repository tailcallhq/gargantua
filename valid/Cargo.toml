[package]
name = "valid"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true


[dev-dependencies]
pretty_assertions.workspace = true
serde.workspace = true
serde_json.workspace = true
stripmargin = "0.1.1"
serde_path_to_error = "0.1.16"
