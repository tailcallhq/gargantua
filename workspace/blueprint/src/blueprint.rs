use std::collections::BTreeSet;

use async_graphql_parser::types::ServiceDocument;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::error::Error;
use crate::index::Index;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Blueprint {
    pub definitions: Vec<Definition>,
    pub schema: SchemaDefinition,
    pub directives: Vec<DirectiveDefinition>,
    pub join_graphs: Vec<JoinGraph>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct Graph(String);
impl Graph {
    pub fn new<A: AsRef<str>>(name: A) -> Self {
        Graph(name.as_ref().to_string())
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub struct JoinGraph {
    pub name: Graph,
    pub url: url::Url,
}

impl Blueprint {
    pub fn parse_doc(doc: ServiceDocument) -> Blueprint {
        super::build::parse(doc)
    }

    pub fn parse(schema: &str) -> Result<Blueprint, Error> {
        let document = async_graphql_parser::parse_schema(schema)?;
        Ok(Blueprint::parse_doc(document))
    }

    pub fn to_index(&self) -> Index {
        Index::from(self)
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Definition {
    Interface(InterfaceTypeDefinition),
    Object(ObjectTypeDefinition),
    InputObject(InputObjectTypeDefinition),
    Scalar(ScalarTypeDefinition),
    Enum(EnumTypeDefinition),
    Union(UnionTypeDefinition),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InterfaceTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub join_types: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub implements: BTreeSet<String>,
    pub join_types: Vec<JoinType>,
    pub join_implements: Vec<JoinImplements>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<InputFieldDefinition>,
    pub description: Option<String>,
    pub join_types: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub enum_values: Vec<EnumValueDefinition>,
    pub join_types: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EnumValueDefinition {
    pub description: Option<String>,
    pub name: String,
    pub directives: Vec<Directive>,
    pub join_enums: Vec<JoinEnum>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SchemaDefinition {
    pub query: Option<String>,
    pub mutation: Option<String>,
    pub subscription: Option<String>,
    pub directives: Vec<Directive>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InputFieldDefinition {
    pub name: String,
    pub of_type: Type,
    pub default_value: Option<Value>,
    pub description: Option<String>,
    pub join_fields: Vec<JoinField>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub name: String,
    pub args: Vec<InputFieldDefinition>,
    pub of_type: Type,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_fields: Vec<JoinField>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Directive {
    pub name: String,
    pub arguments: Value,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DirectiveDefinition {
    pub name: String,
    pub description: Option<String>,
    pub arguments: Vec<InputFieldDefinition>,
    pub repeatable: bool,
    pub locations: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ScalarTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub join_types: Vec<JoinType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UnionTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub types: BTreeSet<String>,
    pub join_types: Vec<JoinType>,
    pub join_unions: Vec<JoinUnion>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinType {
    pub graph: Graph,
    pub key: Option<String>,
    #[serde(default = "default_false")]
    pub extension: bool,
    #[serde(default = "default_true")]
    pub resolvable: bool,
    #[serde(default = "default_false")]
    pub is_interface_object: bool,
}

fn default_true() -> bool {
    true
}
fn default_false() -> bool {
    true
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinEnum {
    pub graph: Graph,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct JoinField {
    pub graph: Option<Graph>,
    pub requires: Option<String>,
    pub provides: Option<String>,
    pub r#type: Option<String>,
    pub external: Option<bool>,
    pub r#override: Option<String>,
    pub used_overridden: Option<bool>,
}

impl JoinField {
    pub fn new(g: Graph) -> Self {
        Self { graph: Some(g), ..Default::default() }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinImplements {
    pub graph: Graph,
    pub interface: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct JoinUnion {
    pub graph: Graph,
    pub member: String,
}

/// Type to represent GraphQL type usage with modifiers
/// [spec](https://spec.graphql.org/October2021/#sec-Wrapping-Types)
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Type {
    Named {
        /// Name of the type
        name: String,
        /// Flag to indicate the type is required.
        required: bool,
    },
    List {
        /// Type is a list
        of_type: Box<Type>,
        /// Flag to indicate the type is required.
        non_null: bool,
    },
}

impl Type {
    pub fn as_type_str(&self) -> String {
        match self {
            Type::Named { name, .. } => name.to_string(),
            Type::List { of_type, .. } => of_type.as_type_str(),
        }
    }
}
