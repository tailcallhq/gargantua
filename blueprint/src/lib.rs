mod blueprint;
mod build;

pub use blueprint::*;
pub use build::parse;
pub use {async_graphql_parser, async_graphql_value, url};

#[cfg(test)]
mod tests {
    use valid::Validator;

    use super::*;

    #[test]
    fn test_parse() {
        let graphql = resource::resource_str!("../examples/router.graphql");
        let document = async_graphql_parser::parse_schema(graphql).unwrap();
        let blueprint = parse(document)
            .map(|b| serde_json::to_string_pretty(&b).unwrap())
            .to_result()
            .unwrap();
        insta::assert_snapshot!(blueprint);
    }
}
