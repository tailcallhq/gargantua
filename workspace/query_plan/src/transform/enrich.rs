use std::marker::PhantomData;
use std::rc::Rc;

use blueprint::{Index, QueryField};
use valid::{Transform, Valid, Validator};

use crate::{Fetch, QueryPlan, SelectionSet, TypeName};

pub struct Enrich<Value> {
    index: Rc<Index>,
    _marker: PhantomData<Value>,
}

impl<Value: Clone> Enrich<Value> {
    pub fn new(index: Rc<Index>) -> Self {
        Self { index, _marker: PhantomData }
    }

    fn iter_sel(
        &self,
        selection: SelectionSet<Value>,
        parent_type: &str,
    ) -> Valid<SelectionSet<Value>, String> {
        // this field belongs to container_type, so we if want to get this field
        let type_def = match self.index.get_object_type_definition(parent_type) {
            Some(type_def) => type_def,
            None => {
                return Valid::fail(format!(
                    "type definition not found for type '{}' ",
                    parent_type
                ));
            }
        };

        Valid::from_iter(selection.into_vec(), |field| {
            let field_def = match self.index.get_field(parent_type, &field.name) {
                Some(QueryField::Field((def, _))) => def,
                _ => {
                    return Valid::fail(format!(
                        "field definition not found for field '{}' in type '{}' ",
                        field.name, parent_type
                    ));
                }
            };

            let mut field = if field_def.join_fields.is_empty() {
                // if field doesn't have @join__field directive, then
                // we need to figure out from where this field can be queried

                // 1. this field can be queried form the @join__type -> wherein the key is same
                //    as this field.
                // 2. this field can be queried from the @join__type directive's graph where key
                //    is none.
                let graphs = type_def
                    .join_types
                    .iter()
                    .filter_map(|jt| {
                        if jt.key.is_none() || jt.key.as_ref().map_or(false, |_k| false) { // TODO: fix that
                            Some(jt.graph.clone())
                        } else {
                            None
                        }
                    })
                    .collect::<Vec<_>>();

                field.graph(graphs)
            } else {
                field.join_field(field_def.join_fields.clone())
            };

            let type_name = field_def.of_type.as_type_str();

            field = field
                .field_type(Some(TypeName::new(type_name.to_string())))
                .parent_type(Some(TypeName::new(parent_type.to_string())));

            if !field.selections.is_empty() {
                let selection = field.selections.clone();
                self.iter_sel(selection, &type_name)
                    .map(|selection_set| field.selections(selection_set))
            } else {
                Valid::succeed(field)
            }
        })
        .map(|fields| SelectionSet::new(fields))
    }

    fn iter(
        &self,
        query: QueryPlan<Value>,
        container_type: &str,
    ) -> Valid<QueryPlan<Value>, String> {
        match query {
            QueryPlan::Fetch(Fetch {
                name,
                arguments,
                variables,
                directives,
                selection_set,
                representations,
                type_name,
                service,
            }) => self
                .iter_sel(selection_set, container_type)
                .map(|selection_set| {
                    QueryPlan::Fetch(Fetch {
                        name,
                        arguments,
                        variables,
                        directives,
                        selection_set,
                        representations,
                        type_name,
                        service,
                    })
                }),
            QueryPlan::Flatten { select, plan } => self
                .iter(*plan, container_type)
                .map(|plan| QueryPlan::Flatten { select, plan: Box::new(plan) }),

            QueryPlan::Parallel(plans) => {
                Valid::from_iter(plans, |plan| self.iter(plan, container_type))
                    .map(|plans| QueryPlan::Parallel(plans))
            }

            QueryPlan::Sequence(plans) => {
                Valid::from_iter(plans, |plan| self.iter(plan, container_type))
                    .map(|plans| QueryPlan::Sequence(plans))
            }
        }
    }
}

impl<Value: Clone> Transform for Enrich<Value> {
    type Value = QueryPlan<Value>;
    type Error = String;

    fn transform(&self, value: Self::Value) -> valid::Valid<Self::Value, Self::Error> {
        Valid::from_option(
            self.index.get_query(),
            "Root operation for `query` is not defined".to_string(),
        )
        .and_then(|root_type| self.iter(value, root_type))
    }
}

#[cfg(test)]
mod test {
    use blueprint::Blueprint;

    use super::*;

    fn setup(graphql: &str) -> Index {
        let document = async_graphql_parser::parse_schema(graphql).unwrap();
        Blueprint::parse_doc(document).to_index()
    }

    #[test]
    fn test_enricher_supergraph_1() {
        let query = "query { topProducts { productName: name reviews { body } reviews { id } } }";
        let index = setup(include_str!(
            "../../../blueprint/src/fixtures/router.graphql"
        ));
        let qp = QueryPlan::try_new(query).unwrap();

        let enriched_selection_set = Enrich::new(Rc::new(index))
            .transform(qp)
            .to_result()
            .unwrap();

        insta::assert_debug_snapshot!(enriched_selection_set)
    }
}
