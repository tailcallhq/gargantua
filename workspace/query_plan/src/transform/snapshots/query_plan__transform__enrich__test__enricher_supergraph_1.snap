---
source: workspace/query_plan/src/transform/enrich.rs
expression: enriched_selection_set
---
Parallel(
    [
        Fetch(
            Fetch {
                name: None,
                arguments: [],
                variables: [],
                directives: [],
                selection_set: SelectionSet(
                    [
                        Field {
                            name: "topProducts",
                            alias: None,
                            selections: SelectionSet(
                                [
                                    Field {
                                        name: "name",
                                        alias: Some(
                                            "productName",
                                        ),
                                        selections: SelectionSet(
                                            [],
                                        ),
                                        arguments: [],
                                        directives: [],
                                        is_hidden: false,
                                        graph: [],
                                        join_field: [
                                            JoinFieldParsed {
                                                graph: Some(
                                                    Graph(
                                                        "PRODUCTS",
                                                    ),
                                                ),
                                                requires: None,
                                                provides: None,
                                                type: None,
                                                external: None,
                                                override: None,
                                                used_overridden: None,
                                            },
                                        ],
                                        field_type: Some(
                                            TypeName(
                                                "String",
                                            ),
                                        ),
                                        parent_type: Some(
                                            TypeName(
                                                "Product",
                                            ),
                                        ),
                                    },
                                    Field {
                                        name: "reviews",
                                        alias: None,
                                        selections: SelectionSet(
                                            [
                                                Field {
                                                    name: "body",
                                                    alias: None,
                                                    selections: SelectionSet(
                                                        [],
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    is_hidden: false,
                                                    graph: [
                                                        Graph(
                                                            "REVIEWS",
                                                        ),
                                                    ],
                                                    join_field: [],
                                                    field_type: Some(
                                                        TypeName(
                                                            "String",
                                                        ),
                                                    ),
                                                    parent_type: Some(
                                                        TypeName(
                                                            "Review",
                                                        ),
                                                    ),
                                                },
                                            ],
                                        ),
                                        arguments: [],
                                        directives: [],
                                        is_hidden: false,
                                        graph: [],
                                        join_field: [
                                            JoinFieldParsed {
                                                graph: Some(
                                                    Graph(
                                                        "REVIEWS",
                                                    ),
                                                ),
                                                requires: None,
                                                provides: None,
                                                type: None,
                                                external: None,
                                                override: None,
                                                used_overridden: None,
                                            },
                                        ],
                                        field_type: Some(
                                            TypeName(
                                                "Review",
                                            ),
                                        ),
                                        parent_type: Some(
                                            TypeName(
                                                "Product",
                                            ),
                                        ),
                                    },
                                    Field {
                                        name: "reviews",
                                        alias: None,
                                        selections: SelectionSet(
                                            [
                                                Field {
                                                    name: "id",
                                                    alias: None,
                                                    selections: SelectionSet(
                                                        [],
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    is_hidden: false,
                                                    graph: [
                                                        Graph(
                                                            "REVIEWS",
                                                        ),
                                                    ],
                                                    join_field: [],
                                                    field_type: Some(
                                                        TypeName(
                                                            "ID",
                                                        ),
                                                    ),
                                                    parent_type: Some(
                                                        TypeName(
                                                            "Review",
                                                        ),
                                                    ),
                                                },
                                            ],
                                        ),
                                        arguments: [],
                                        directives: [],
                                        is_hidden: false,
                                        graph: [],
                                        join_field: [
                                            JoinFieldParsed {
                                                graph: Some(
                                                    Graph(
                                                        "REVIEWS",
                                                    ),
                                                ),
                                                requires: None,
                                                provides: None,
                                                type: None,
                                                external: None,
                                                override: None,
                                                used_overridden: None,
                                            },
                                        ],
                                        field_type: Some(
                                            TypeName(
                                                "Review",
                                            ),
                                        ),
                                        parent_type: Some(
                                            TypeName(
                                                "Product",
                                            ),
                                        ),
                                    },
                                ],
                            ),
                            arguments: [],
                            directives: [],
                            is_hidden: false,
                            graph: [],
                            join_field: [
                                JoinFieldParsed {
                                    graph: Some(
                                        Graph(
                                            "PRODUCTS",
                                        ),
                                    ),
                                    requires: None,
                                    provides: None,
                                    type: None,
                                    external: None,
                                    override: None,
                                    used_overridden: None,
                                },
                            ],
                            field_type: Some(
                                TypeName(
                                    "Product",
                                ),
                            ),
                            parent_type: Some(
                                TypeName(
                                    "Query",
                                ),
                            ),
                        },
                    ],
                ),
                representations: None,
                type_name: TypeName(
                    "query",
                ),
                service: None,
            },
        ),
    ],
)
